### Importing Libraries

In [29]:
import mysql.connector
import pandas as pd
from scipy import stats

### Connecting to SQL

In [3]:
fortuneDatabase = mysql.connector.connect(
    host = "localhost",
    user = "root",
    password = "IshuLogi@9702",
    database = "fortune_companies_db"
)

### Fetching Data from SQL

In [4]:
cursor_func = fortuneDatabase.cursor()
cursor_func.execute("SELECT * FROM fortune_2023_and_2024")
result = cursor_func.fetchall()
# result

### Fetching Column Names

In [5]:
cursor_func.description

[('2023_Rank', 8, None, None, None, None, 1, 32768, 63),
 ('2023_Company', 252, None, None, None, None, 1, 16, 255),
 ('2023_Sector', 252, None, None, None, None, 1, 16, 255),
 ('2023_Industry', 252, None, None, None, None, 1, 16, 255),
 ('2023_Profitable', 252, None, None, None, None, 1, 16, 255),
 ('2023_Founder_is_CEO', 252, None, None, None, None, 1, 16, 255),
 ('2023_FemaleCEO', 252, None, None, None, None, 1, 16, 255),
 ('2023_Growth_in_Jobs', 252, None, None, None, None, 1, 16, 255),
 ('2023_Change_in_Rank', 5, None, None, None, None, 1, 32768, 63),
 ('2023_Gained_in_Rank', 252, None, None, None, None, 1, 16, 255),
 ('2023_Dropped_in_Rank', 252, None, None, None, None, 1, 16, 255),
 ('2023_Newcomer_to_the_Fortune500', 252, None, None, None, None, 1, 16, 255),
 ('2023_Global500', 252, None, None, None, None, 1, 16, 255),
 ('2023_Best_Companies', 252, None, None, None, None, 1, 16, 255),
 ('2023_Number_of_employees', 8, None, None, None, None, 1, 32768, 63),
 ('2023_MarketCap_Marc

### Creating Dataframe

In [6]:
fortuneColumnNames = [i[0] for i in cursor_func.description]

fortuneDataframe = pd.DataFrame(result, columns = fortuneColumnNames)
fortuneDataframe.head(3)

2023_Rank            2023_Company  2023_Sector  \
0      116.0                      3M    Chemicals   
1      458.0  A-Mark Precious Metals  Wholesalers   
2      780.0              A.O. Smith  Industrials   

                    2023_Industry 2023_Profitable 2023_Founder_is_CEO  \
0                       Chemicals             yes                  no   
1        Wholesalers: Diversified             yes                  no   
2  Electronics, Electrical Equip.             yes                  no   

  2023_FemaleCEO 2023_Growth_in_Jobs  2023_Change_in_Rank 2023_Gained_in_Rank  \
0             no                  no                -14.0                  no   
1             no                 yes                -12.0                  no   
2             no                  no                -34.0                  no   

   ... 2024_RevenuePercentChange 2024_Profits_M 2024_ProfitsPercentChange  \
0  ...                      -4.5        -6995.0                    -221.1   
1  ...                      13.8          156.4                      18.0   
2  ...                       2.6          556.6                     136.1   

  2024_Assets_M            2024_CEO  2024_HeadquartersCity  \
0       50580.0    William M. Brown               St. Paul   
1        1545.6  Gregory N. Roberts             El Segundo   
2        3213.9    Kevin J. Wheeler              Milwaukee   

   2024_HeadquartersState  2024_CompanyType  2024_MarketCap_Updated_M  \
0               Minnesota            Public                   56841.0   
1              California            Public                     701.0   
2               Wisconsin            Public                   13156.0   

   2024_Year  
0     2024.0  
1     2024.0  
2     2024.0  

[3 rows x 55 columns]

### Concatenating Columns

In [7]:
fortuneDataframe['HeadquartersState'] = fortuneDataframe.apply(
    lambda row: row['2024_HeadquartersState'] 
    if pd.notnull(row['2024_HeadquartersState']) and row['2024_HeadquartersState'] != ''
    else row['2023_HeadquartersState'], axis = 1
)

### Replacing Headquarters State Column Values

In [8]:
renameHeadquartersState = {
    'OR' : 'Oregon',
    'IL' : 'Illinois',
    'WA' : 'Washington',
    'AZ' : 'Arizona',
    'CO' : 'Colorado',
    'FL' : 'Florida',
    'GA' : 'Georgia',
    'NJ' : 'New Jersey',
    'MO' : 'Missouri',
    'VA' : 'Virginia',
    'UT' : 'Utah',
    'CA' : 'California',
    'TN' : 'Tennessee',
    'MI' : 'Michigan',
    'IN' : 'Indiana',
    'DE' : 'Delaware',
    'NY' : 'New York',
    'TX' : 'Texas',
    'OH' : 'Ohio',
    'PA' : 'Pennsylvania',
    'OK' : 'Oklahoma',
    'WI' : 'Wisconsin',
    'CT' : 'Connecticut',
    'NC' : 'North Carolina',
    'MN' : 'Minnesota',
    'SD' : 'South Dakota'
}

fortuneDataframe['HeadquartersState'] = fortuneDataframe['HeadquartersState'].str.strip()

fortuneDataframe['HeadquartersState'] = fortuneDataframe['HeadquartersState'].replace(renameHeadquartersState)
# fortuneDataframe

### Check Unique Values of Headquarters State

In [9]:
print(fortuneDataframe['HeadquartersState'].unique())

['Minnesota' 'California' 'Wisconsin' 'Illinois' 'Ohio' 'New York' 'Texas'
 'Tennessee' 'Georgia' 'Pennsylvania' 'Florida' 'North Carolina'
 'Virginia' 'Massachusetts' 'Washington' 'Idaho' 'Arizona' 'Oklahoma'
 'Indiana' 'Michigan' 'Kansas' 'Missouri' 'New Jersey' 'Rhode Island'
 'Connecticut' 'Colorado' 'Arkansas' 'Delaware' 'Nebraska' 'South Dakota'
 'Nevada' 'Kentucky' 'District Of Columbia' 'Iowa' 'Maryland' 'Oregon'
 'South Carolina' 'Alabama' 'Louisiana' 'Hawaii' 'Maine' 'New Hampshire'
 'North Dakota' 'Puerto Rico' 'Utah' 'Mississippi' 'Vermont' 'Montana']


### Check Null Values

In [10]:
fortuneDataframe.isnull().sum()

2023_Rank                              75
2023_Company                           75
2023_Sector                            75
2023_Industry                          75
2023_Profitable                        75
2023_Founder_is_CEO                    75
2023_FemaleCEO                         75
2023_Growth_in_Jobs                    82
2023_Change_in_Rank                   146
2023_Gained_in_Rank                   146
2023_Dropped_in_Rank                  146
2023_Newcomer_to_the_Fortune500       575
2023_Global500                         75
2023_Best_Companies                    75
2023_Number_of_employees               75
2023_MarketCap_March31_M              116
2023_Revenues_M                        75
2023_RevenuePercentChange              84
2023_Profits_M                         75
2023_ProfitsPercentChange             209
2023_Assets_M                          75
2023_CEO                               76
2023_HeadquartersCity                  75
2023_HeadquartersState            

### Concatenating Columns

In [11]:
fortuneDataframe['Company'] = fortuneDataframe.apply(
    lambda row: row['2024_Company']
    if pd.notnull(row['2024_Company']) and row['2024_Company'] != ''
    else row['2023_Company'], axis = 1
)

In [12]:
fortuneDataframe['Sector'] = fortuneDataframe.apply(
    lambda row: row['2024_Sector']
    if pd.notnull(row['2024_Sector']) and row['2024_Sector'] != ''
    else row['2023_Sector'], axis = 1
)

In [13]:
fortuneDataframe['Industry'] = fortuneDataframe.apply(
    lambda row: row['2024_Industry']
    if pd.notnull(row['2024_Industry']) and row['2024_Industry'] != ''
    else row['2023_Industry'], axis = 1
)

In [14]:
fortuneDataframe['CompanyType'] = fortuneDataframe.apply(
    lambda row: row['2024_CompanyType']
    if pd.notnull(row['2024_CompanyType']) and row['2024_CompanyType'] != ''
    else row['2023_CompanyType'], axis = 1
)

### Dropping Columns

In [15]:

fortuneDataframe = fortuneDataframe.drop(['2023_HeadquartersCity', '2023_HeadquartersState', '2024_HeadquartersCity', '2024_HeadquartersState', 
                                          '2023_Company', '2024_Company', '2023_Sector', '2024_Sector', '2023_Industry', '2024_Industry', 
                                          '2023_CompanyType', '2024_CompanyType'], axis = 1)

In [16]:

fortuneDataframe[['2023_Rank', '2024_Rank', '2023_Year', '2024_Year', '2023_Number_of_employees', '2024_Number_of_employees', 
                  '2023_MarketCap_Updated_M', '2024_MarketCap_Updated_M', '2023_Change_in_Rank', 
                  '2024_Change_in_Rank']] = fortuneDataframe[['2023_Rank', '2024_Rank', '2023_Year', 
                                                              '2024_Year', '2023_Number_of_employees', '2024_Number_of_employees', 
                                                              '2023_MarketCap_Updated_M', '2024_MarketCap_Updated_M',  
                                                              '2023_Change_in_Rank', '2024_Change_in_Rank']].fillna(0).astype('int')

### Formatting Numerical Values

In [17]:

def format_float(value):
    if pd.isna(value):
        return value
    roundOffVal = round(value, 0)
    return int(roundOffVal) if roundOffVal.is_integer() else roundOffVal


columns_to_format = ["2023_MarketCap_March31_M", "2023_Revenues_M", "2023_RevenuePercentChange", "2023_Profits_M", "2023_ProfitsPercentChange", 
                     "2023_Assets_M", "2024_MarketCap_March28_M", "2024_Revenues_M", "2024_RevenuePercentChange", 
                     "2024_Profits_M", "2024_ProfitsPercentChange", "2024_Assets_M"]

for columns in columns_to_format:
    fortuneDataframe[columns] = fortuneDataframe[columns].apply(format_float)

fortuneDataframe[["2023_MarketCap_March31_M", "2023_Revenues_M", "2023_RevenuePercentChange", "2023_Profits_M", "2023_ProfitsPercentChange", 
                     "2023_Assets_M", "2024_MarketCap_March28_M", "2024_Revenues_M", "2024_RevenuePercentChange", 
                     "2024_Profits_M", "2024_ProfitsPercentChange", "2024_Assets_M"]] = fortuneDataframe[["2023_MarketCap_March31_M", "2023_Revenues_M", "2023_RevenuePercentChange", "2023_Profits_M", "2023_ProfitsPercentChange", 
                     "2023_Assets_M", "2024_MarketCap_March28_M", "2024_Revenues_M", "2024_RevenuePercentChange", 
                     "2024_Profits_M", "2024_ProfitsPercentChange", "2024_Assets_M"]].fillna(0).astype('int')


### Rearranging the Columns

In [19]:
col_rearrange = fortuneDataframe.columns.to_list()

first_columns = ['Company', 'Sector', 'Industry', 'CompanyType', 'HeadquartersState']

new_column_order = [col for col in col_rearrange if col not in first_columns]

fortuneColumnRearrange = first_columns + new_column_order
fortuneDataframe = fortuneDataframe[fortuneColumnRearrange]

fortuneDataframe

Company               Sector  \
0                          3M            Chemicals   
1      A-Mark Precious Metals          Wholesalers   
2                  A.O. Smith          Industrials   
3         Abbott Laboratories          Health Care   
4                      AbbVie          Health Care   
...                       ...                  ...   
1070                   Verisk    Business Services   
1071                      WEX    Business Services   
1072       Wintrust Financial           Financials   
1073                 Woodward  Aerospace & Defense   
1074  Worthington Enterprises            Materials   

                            Industry CompanyType HeadquartersState  2023_Rank  \
0                          Chemicals      Public         Minnesota        116   
1           Wholesalers: Diversified      Public        California        458   
2     Electronics, Electrical Equip.      Public         Wisconsin        780   
3     Medical Products and Equipment      Public          Illinois         99   
4                    Pharmaceuticals      Public          Illinois         73   
...                              ...         ...               ...        ...   
1070         Financial Data Services      Public        New Jersey          0   
1071         Financial Data Services      Public             Maine          0   
1072                Commercial Banks      Public          Illinois          0   
1073             Aerospace & Defense      Public          Colorado          0   
1074                          Metals      Public              Ohio          0   

     2023_Profitable 2023_Founder_is_CEO 2023_FemaleCEO 2023_Growth_in_Jobs  \
0                yes                  no             no                  no   
1                yes                  no             no                 yes   
2                yes                  no             no                  no   
3                yes                  no             no                 yes   
4                yes                  no             no                  no   
...              ...                 ...            ...                 ...   
1070            None                None           None                None   
1071            None                None           None                None   
1072            None                None           None                None   
1073            None                None           None                None   
1074            None                None           None                None   

      ...  2024_Number_of_employees 2024_MarketCap_March28_M 2024_Revenues_M  \
0     ...                     85000                    58695           32681   
1     ...                       421                      701            9287   
2     ...                     12000                    13156            3853   
3     ...                    114000                   197221           40109   
4     ...                     50000                   322435           54318   
...   ...                       ...                      ...             ...   
1070  ...                      7500                    33801            2728   
1071  ...                      7200                     9913            2548   
1072  ...                      5521                     6417            3327   
1073  ...                      8800                     9293            2915   
1074  ...                      8200                     3111            4916   

     2024_RevenuePercentChange 2024_Profits_M 2024_ProfitsPercentChange  \
0                           -4          -6995                      -221   
1                           14            156                        18   
2                            3            557                       136   
3                           -8           5723                       -18   
4                           -6           4863                       -59   
...                        ...          

In [20]:
fortuneDataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1075 entries, 0 to 1074
Data columns (total 48 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Company                             1075 non-null   object
 1   Sector                              1075 non-null   object
 2   Industry                            1075 non-null   object
 3   CompanyType                         1075 non-null   object
 4   HeadquartersState                   1075 non-null   object
 5   2023_Rank                           1075 non-null   int32 
 6   2023_Profitable                     1000 non-null   object
 7   2023_Founder_is_CEO                 1000 non-null   object
 8   2023_FemaleCEO                      1000 non-null   object
 9   2023_Growth_in_Jobs                 993 non-null    object
 10  2023_Change_in_Rank                 1075 non-null   int32 
 11  2023_Gained_in_Rank                 929 non-null    obje

In [34]:
# fortuneDataframe.isnull().sum()

In [35]:
# fortuneDataframe.to_csv("C:/Users/Logiya Vidhyapathy/Documents/KGISL Data Science/Project Files and Documents/MileStone 2 Project/fortune.csv", index = False)

### Statistical Analysis

### Paired T Test

In [33]:
profit_2023 = fortuneDataframe['2023_Profits_M']
profit_2024 = fortuneDataframe['2024_Profits_M']

t_statistic, p_value = stats.ttest_rel(profit_2023, profit_2024)

print(f"t_statistic: {round(t_statistic, 2)}")
print(f"p_value: {round(p_value, 2)}")

t_statistic: -0.92
p_value: 0.36


- __Null Hypothesis (H₀):__ There is no significant difference between 2023_Profits and 2024_Profits.
- __Alternate Hypothesis (H₁):__ There is a significant difference between 2023_Profits and 2024_Profits.

Therefore, we __reject the Null Hypothesis (0.36 > 0.05)__ (i.e. we fail to reject the null hypothesis) 